src/main/java/de/tum/in/www1/artemis/service/ParticipationVcsAccessTokenService.java

```diff
@@ -0,0 +1,122 @@
+            var participationsWithTokens = existingTokens.stream().map(ParticipationVCSAccessToken::getParticipation).toList();
+            log.info("Create missing VcsAccessTokens for participationIds: {}", participations);
+            List<ParticipationVCSAccessToken> vcsAccessTokens = new ArrayList<>();
+
+            for (ProgrammingExerciseStudentParticipation participation : participations) {
+                if (!participationsWithTokens.contains(participation) && participation.getParticipant() instanceof User) {
+                    var participationVCSAccessToken = new ParticipationVCSAccessToken();
+                    participationVCSAccessToken.setUser((User) participation.getParticipant());
+                    participationVCSAccessToken.setParticipation(participation);
+                    participationVCSAccessToken.setVcsAccessToken(LocalVCPersonalAccessTokenManagementService.generateSecureVCSAccessToken());
+                    vcsAccessTokens.add(participationVCSAccessToken);
+                }
+            }
+
+            log.info("Successfully generated {} missing VcsAccessTokens", vcsAccessTokens);
+            participationVcsAccessTokenRepository.saveAll(vcsAccessTokens);
+            log.info("Successfully saved missing VcsAccessTokens");
+        }
+        catch (Exception e) {
+            log.error("Error creating missing VCS access tokens: {}", e.getMessage());
+        }
+    }
+
+    /**
+     * Retrieves the participationVCSAccessToken for a User,Participation pair if it exists
+     *
+     * @param userId          the user's id which is owner of the token
+     * @param participationId the participation's id which the token belongs to
+     * @return an Optional participationVCSAccessToken,
+     */
+    public Optional<ParticipationVCSAccessToken> findByUserIdAndParticipationId(Long userId, Long participationId) {
+        return participationVcsAccessTokenRepository.findByUserIdAndParticipationId(userId, participationId);
+    }
+
+    /**
+     * Deletes the token connected to a participation
+     *
+     * @param participationId the participation id for which the token should get deleted
+     */
+    public void deleteByParticipationId(long participationId) {
+        participationVcsAccessTokenRepository.deleteByParticipation_id(participationId);
+    }
+
+    public void deleteAllByParticipations(List<ProgrammingExerciseStudentParticipation> participations) {
+        for (ProgrammingExerciseStudentParticipation participation : participations) {
+            participationVcsAccessTokenRepository.deleteByParticipation_id(participation.getId());
+        }
+    }
+
+    /**
+     * Create missing participationVcsAccessTokens at application start
+     */
+    @EventListener(ApplicationReadyEvent.class)
+    public void createMissingParticipationVcsAccessToken() {
```

Comment:
Do we need this for every Application start? Otherwise we should move this to a JavaMigration like `MigrationEntry20240614_140000`

Reply:
This would also work and I thought about doing it this way, [see this old commit](https://github.com/ls1intum/Artemis/pull/8929/commits/cddfaa29f5faf6ef2a57e20201a8894938fb243a#diff-eb5161fdeb9dc37f8cac73a4c49b19635a085e2b949244bfe11dd2330743755c), but this wouldn't help if due to some other reasons (e.g. migration from gitlab to localvc maybe), some participations don't have a token associated with them

Reply:
I decided to create tokens on demand, if they don't already exist. This removes the overhead of a migration or a creation at startup (would probably not be the best, as for many old and unused participations, it would be a waste of resources to generate and store tokens for all of them. )